examples/wave.cxx

```diff
@@ -0,0 +1,372 @@
+// LLC, THE U.S. DEPARTMENT OF ENERGY OR CONTRIBUTORS BE LIABLE FOR ANY
+// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+// OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+// IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include <cstdlib>
+#include <iostream>
+#include <cmath>
+
+#include "RAJA/RAJA.hpp"
+#include "RAJA/util/defines.hpp"
+
+#define sr 2 //Stencil Radius
+#define PI 3.14159265359
+#define CUDA_BLOCK_SIZE 16
+
+
+typedef struct grid{
+  double ox,dx;
+  int nx;
+}grid_s;
+
+
+template <typename T>
+T* allocate(RAJA::Index_type size) {
+
+  T* ptr;
+#if defined(RAJA_ENABLE_CUDA)
+  cudaMallocManaged((void**)&ptr, sizeof(T)*size,cudaMemAttachGlobal);
+#else
+  ptr = new T[size];
+#endif
+  return ptr;
+}
+
+template <typename T>
+void deallocate(T* &ptr) {
+  if (ptr) {
+#if defined(RAJA_ENABLE_CUDA)
+    cudaFree(ptr);
+#else
+    delete[] ptr;
+#endif
+    ptr = nullptr;
+  }
+}
+
+
+
+struct DefaultCPP{};
+
+//Generic RAJA implementation for CPU
+template <typename Policy1, typename Policy2>
+struct Wave {
+  template <typename T>  
+    void operator()(T* P1, T* P2, int nx, double ct) {
+
+    RAJA::forallN< RAJA::NestedPolicy<
+    RAJA::ExecList<RAJA::seq_exec,RAJA::seq_exec >>>(
+    RAJA::RangeSegment(0, nx),
+    RAJA::RangeSegment(0, nx),
+    [=](int ty, int tx) {
+    
+    double coeff[5] = {-1.0/12.0,4.0/3.0,-5.0/2.0,4.0/3.0,-1.0/12.0}; //fourth order scheme
+    
+    // loop over points        
+    const int id = tx + ty*nx;
+    double P_old  = P1[id]; 
+    double P_curr = P2[id]; 
+    
+    //Compute laplacian
+    double lap = 0.0;
+    
+    for(int r=-sr; r<=sr; ++r){
+      const int xi  = (tx+r+nx)%nx;
+      const int idx = xi + nx*ty;
+      lap += coeff[r+sr]*P2[idx]; 
+      
+      const int yi  = (ty+r+nx)%nx;
+      const int idy = tx + nx*yi;
+      lap += coeff[r+sr]*P2[idy];
+    }
+    
+    //finite difference stencil
+    P1[id] = 2*P_curr - P_old + ct*lap;
+          
+    }); 
+            
+  }
+};
+
+
+// Generic CPP implementation
+template <>
+struct Wave<DefaultCPP, DefaultCPP> {
+  template <typename T>
+  void operator()(T* P1, T* P2, int nx, double ct) {
+
+    double coeff[5] = {-1.0/12.0,4.0/3.0,-5.0/2.0,4.0/3.0,-1.0/12.0}; //fourth order scheme
+    
+    // loop over points
+    for(int ty=0; ty<nx; ++ty) {
+      for(int tx=0; tx<nx; ++tx) {
+        
+        const int id = tx + ty*nx;
+        T P_old  = P1[id]; 
+        T P_curr = P2[id]; 
+        
+        //Compute laplacian
+        T lap = 0.0;
+        
+        for(int r=-sr; r<=sr; ++r) {
```

Comment:
I don't know if we want to try to show anything beyond a simple for-loop, but

`for (auto r : RAJA::RangeSegment(-sr, sr + 1))` would be pretty cool to highlight :)